File: server/pkg/warrant/scope.go
package warrant

import (
	"strings"
)

// ScopeEntry is a parsed "resource:action" permission pair.
type ScopeEntry struct {
	Resource string
	Action   string
}

// ParseScope parses a "resource:action" string.
// Returns (entry, true) on success, (zero, false) if the format is invalid.
func ParseScope(s string) (ScopeEntry, bool) {
	parts := strings.SplitN(s, ":", 2)
	if len(parts) != 2 || parts[0] == "" || parts[1] == "" {
		return ScopeEntry{}, false
	}
	return ScopeEntry{Resource: parts[0], Action: parts[1]}, true
}

// EntryCovers reports whether parent covers child.
// A wildcard ("*") in either position matches any value.
func EntryCovers(parent, child ScopeEntry) bool {
	resourceOK := parent.Resource == "*" || parent.Resource == child.Resource
	actionOK := parent.Action == "*" || parent.Action == child.Action
	return resourceOK && actionOK
}

// IsSubset reports whether every entry in child is covered by at least one
// entry in parent. Invalid entries in child are treated as uncoverable.
func IsSubset(parent, child []string) bool {
	for _, ce := range child {
		childEntry, ok := ParseScope(ce)
		if !ok {
			return false
		}
		covered := false
		for _, pe := range parent {
			parentEntry, ok := ParseScope(pe)
			if !ok {
				continue
			}
			if EntryCovers(parentEntry, childEntry) {
				covered = true
				break
			}
		}
		if !covered {
			return false
		}
	}
	return true
}

// NormaliseScope deduplicates scope entries and sorts them for stable output.
func NormaliseScope(scope []string) []string {
	seen := make(map[string]struct{}, len(scope))
	out := make([]string, 0, len(scope))
	for _, s := range scope {
		s = strings.TrimSpace(s)
		if s == "" {
			continue
		}
		if _, dup := seen[s]; dup {
			continue
		}
		seen[s] = struct{}{}
		out = append(out, s)
	}
	// Insertion-order dedup (deterministic enough for token claims).
	return out
}

File: sdks/python/attest/scope.py
"""Scope parsing and subset-checking utilities for Attest."""

from __future__ import annotations


def parse_scope(s: str) -> tuple[str, str] | None:
    """Parse a ``"resource:action"`` string into ``(resource, action)``.

    Returns ``None`` if the string is not in the expected format (i.e. does
    not contain exactly one colon with non-empty parts on both sides).
    """
    parts = s.split(":", 1)
    if len(parts) != 2 or not parts[0] or not parts[1]:
        return None
    return (parts[0], parts[1])


def entry_covers(parent: tuple[str, str], child: tuple[str, str]) -> bool:
    """Return ``True`` if *parent* covers *child*, respecting ``'*'`` wildcards.

    A wildcard in either the resource or action position of the parent matches
    any corresponding value in the child.
    """
    resource_ok = parent[0] == "*" or parent[0] == child[0]
    action_ok = parent[1] == "*" or parent[1] == child[1]
    return resource_ok and action_ok


def is_subset(parent_scope: list[str], child_scope: list[str]) -> bool:
    """Return ``True`` if every entry in *child_scope* is covered by at least
    one entry in *parent_scope*.

    Invalid scope entries in *child_scope* (those that cannot be parsed) are
    treated as uncoverable and cause this function to return ``False``.
    Invalid entries in *parent_scope* are silently skipped.
    """
    for child_entry_str in child_scope:
        child_parsed = parse_scope(child_entry_str)
        if child_parsed is None:
            return False

        covered = False
        for parent_entry_str in parent_scope:
            parent_parsed = parse_scope(parent_entry_str)
            if parent_parsed is None:
                continue
            if entry_covers(parent_parsed, child_parsed):
                covered = True
                break

        if not covered:
            return False

    return True


def normalise_scope(scope: list[str]) -> list[str]:
    """Deduplicate and strip whitespace from scope entries, preserving order.

    Empty strings (after stripping) are dropped.  Duplicate entries (after
    stripping) keep only the first occurrence.
    """
    seen: set[str] = set()
    out: list[str] = []
    for entry in scope:
        stripped = entry.strip()
        if not stripped:
            continue
        if stripped in seen:
            continue
        seen.add(stripped)
        out.append(stripped)
    return out
